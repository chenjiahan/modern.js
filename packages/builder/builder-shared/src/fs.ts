import { CSS_MODULE_REGEX, CSS_REGEX, NODE_MODULES_REGEX } from './constants';

export async function isFileExists(file: string) {
  const { promises, constants } = await import('fs');
  return promises
    .access(file, constants.F_OK)
    .then(() => true)
    .catch(() => false);
}

export const isNodeModulesCss = (path: string) =>
  NODE_MODULES_REGEX.test(path) &&
  CSS_REGEX.test(path) &&
  !CSS_MODULE_REGEX.test(path);
