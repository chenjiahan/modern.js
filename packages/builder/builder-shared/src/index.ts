export * from './types';
export * from './utils';
export * from './logger';
export * from './context';
export * from './createHook';
