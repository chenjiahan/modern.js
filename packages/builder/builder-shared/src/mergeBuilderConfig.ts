import _ from '@modern-js/utils/lodash';

export const mergeBuilderConfig = <T>(config: T, ...sources: T[]): T =>
  _.mergeWith(config, ...sources, (target: unknown, source: unknown) => {
    const pair = [target, source];
    if (pair.some(_.isUndefined)) {
      // fallback to lodash default merge behavior
      return undefined;
    }
    if (pair.some(_.isArray)) {
      return [..._.castArray(target), ..._.castArray(source)];
    }
    // convert function to chained function
    if (pair.some(_.isFunction)) {
      return [target, source];
    }
    // fallback to lodash default merge behavior
    return undefined;
  });
