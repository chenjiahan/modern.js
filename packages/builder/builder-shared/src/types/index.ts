/* eslint-disable @typescript-eslint/no-restricted-imports */
export * from './builder';
export * from './context';
export * from './utils';
export * from './plugin';
export * from './provider';
/* eslint-enable @typescript-eslint/no-restricted-imports */
