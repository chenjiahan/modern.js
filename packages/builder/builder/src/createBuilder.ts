import {
  pick,
  debug,
  applyDefaultBuilderOptions,
  type BuilderInstance,
  type BuilderProvider,
  type CreateBuilderOptions,
} from '@modern-js/builder-shared';
import { createPluginStore } from './createPluginStore';

export async function createBuilder(
  provider: BuilderProvider,
  options: CreateBuilderOptions,
): Promise<BuilderInstance> {
  const builderOptions = applyDefaultBuilderOptions(options);
  const pluginStore = createPluginStore();
  const {
    build,
    publicContext,
    inspectConfig,
    createCompiler,
    startDevServer,
    applyDefaultPlugins,
  } = await provider({
    pluginStore,
    builderOptions,
  });

  debug('add default plugins');
  await applyDefaultPlugins(pluginStore);
  debug('add default plugins done');

  return {
    ...pick(pluginStore, ['addPlugins', 'removePlugins', 'isPluginExists']),
    build,
    createCompiler,
    inspectConfig,
    startDevServer,
    context: publicContext,
  };
}
