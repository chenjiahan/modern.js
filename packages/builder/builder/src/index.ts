import type { BuilderOptions } from '@modern-js/builder-shared';

function applyDefaultOptions(options?: BuilderOptions) {
  return {
    cwd: process.cwd(),
    entry: {},
    target: ['web'],
    configPath: null,
    framework: 'modern-js',
    ...options,
  } as Required<BuilderOptions>;
}

export function createBuilder<P>(provider: P, options: BuilderOptions) {
  const builderOptions = applyDefaultOptions(options);
  return provider.createBuilder()
}
