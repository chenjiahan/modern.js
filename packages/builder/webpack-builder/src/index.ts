export { webpackBuildProvider } from './provider';
export type {
  // Config Types
  DevConfig,
  HtmlConfig,
  OutputConfig,
  SourceConfig,
  BuilderConfig,
  SecurityConfig,
  PerformanceConfig,
  ExperimentsConfig,

  // Third Party Types
  webpack,
  WebpackChain,
  WebpackConfig,
  CSSLoaderOptions,
  CssExtractOptions,
  LessLoaderOptions,
  SassLoaderOptions,
  HTMLPluginOptions,
  StyleLoaderOptions,
  AutoprefixerOptions,
  TerserPluginOptions,
  PostCSSLoaderOptions,
} from './types';
