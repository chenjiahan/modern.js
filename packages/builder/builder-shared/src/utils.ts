import { BuilderOptions } from './types';

export function applyDefaultBuilderOptions(options?: BuilderOptions) {
  return {
    cwd: process.cwd(),
    entry: {},
    target: ['web'],
    configPath: null,
    framework: 'modern-js',
    ...options,
  } as Required<BuilderOptions>;
}
