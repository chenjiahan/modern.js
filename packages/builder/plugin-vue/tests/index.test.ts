import { expect, describe, it } from 'vitest';
import { createStubBuilder } from '@modern-js/builder-webpack-provider/stub';
import { builderPluginVue } from '../src';

describe('plugins/vue', () => {
  it('should add vue-loader and VueLoaderPlugin correctly', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginVue()],
    });
    const config = await builder.unwrapWebpackConfig();

    expect(config).toMatchSnapshot();
  });

  it('should allow to configure vueLoader options', async () => {
    const builder = await createStubBuilder({
      plugins: [
        builderPluginVue({
          vueLoaderOptions: {
            hotReload: false,
          },
        }),
      ],
    });
    const config = await builder.unwrapWebpackConfig();

    expect(config).toMatchSnapshot();
  });
});
