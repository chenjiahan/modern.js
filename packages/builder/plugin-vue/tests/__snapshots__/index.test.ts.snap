// Vitest Snapshot v1

exports[`plugins/vue > should add vue-loader and VueLoaderPlugin correctly 1`] = `
{
  "module": {
    "rules": [
      {
        "test": /\\\\\\.vue\\$/,
        "use": [
          {
            "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/vue-loader/dist/index.js",
            "options": {
              "compilerOptions": {
                "preserveWhitespace": false,
              },
              "experimentalInlineMatchResource": false,
            },
          },
        ],
      },
    ],
  },
  "plugins": [
    Plugin {},
  ],
  "resolve": {
    "extensions": [
      ".vue",
    ],
  },
}
`;

exports[`plugins/vue > should allow to configure vueLoader options 1`] = `
{
  "module": {
    "rules": [
      {
        "test": /\\\\\\.vue\\$/,
        "use": [
          {
            "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/vue-loader/dist/index.js",
            "options": {
              "compilerOptions": {
                "preserveWhitespace": false,
              },
              "experimentalInlineMatchResource": false,
              "hotReload": false,
            },
          },
        ],
      },
    ],
  },
  "plugins": [
    Plugin {},
  ],
  "resolve": {
    "extensions": [
      ".vue",
    ],
  },
}
`;
